base/iterators.jl

```diff
@@ -152,6 +152,9 @@ julia> for (index, value) in enumerate(a)
 """
 enumerate(iter) = Enumerate(iter)
 
+firstindex(e::Enumerate) = firstindex(e.itr)
+getindex(e::Enumerate,i::Int) = (i,getindex(e.itr,i))
```

Comment:
style nit: spaces should typically follow `,`

But I think the definition also needs to be offset by `firstindex`?

And perhaps should have `lastindex` also?

```suggestion
firstindex(e::Enumerate) = firstindex(e.itr)
lastindex(e::Enumerate) = lastindex(e.itr)
getindex(e::Enumerate, i::Int) = (i - firstindex(i) + 1, e.itr[i])
```

Can you also add tests?

Reply:
I think these methods should be restricted to `Enumerate{<:AbstractArray}` [^1]. For example, the proposed `getindex` will not work correctly for something like `Enumerate{String}`:
```julia
julia> enumerate("τεst") |> collect
4-element Vector{Tuple{Int64, Char}}:
 (1, 'τ')
 (2, 'ε')
 (3, 's')
 (4, 't')

julia> enumerate("τεst")[3]
(3, 'ε')
```
Here, indexing produces an element that is not part of the iterator.

[^1]: Of course, one could also implement an `AbstractArray` subtype for which it produces wrong results (see https://github.com/JuliaLang/julia/pull/40772#issuecomment-837997839 above), but it seems less likely to occur in practice.

Reply:
you can already
```julia
julia> "τεst"[3]
'ε': Unicode U+03B5 (category Ll: Letter, lowercase)
```
which is "wrong" but it's legal. Also I don't think there's guarantee against:
> indexing produces an element that is not part of the iterator.

Reply:
> which is "wrong" but it's legal

No, `"τεst"[3] === 'ε'` isn’t wrong. It is the correct and documented behavior.

I agree that there is no guarantee that indexing and iteration produce different values, and I don’t think that’s a problem. Iteration and indexing are two different mechanisms which don’t necessarily mix well.

The problem I see with this PR is that it leads to confusing behavior _because_ it tries to mix indexing and iteration. The motivation for the PR is that 
```julia
Threads.@threads for x in enumerate(foo)
    ...
end
```
currently errors. With this PR, it would no longer error, but it might iterate over something completely different than without the `Threads.@threads`, because `Threads.@threads` rewrites the loop to use indexing instead of iteration. I think erroring in this case is preferable to having the loop iterate over completely different elements than it would without `Threads.@threads`.

Reply:
I just noticed that my `Threads.@threads for x = enumerate("τεst")` example above would still error, since it would try to access `"τεst"[2]` as well (but another thread might still process `(3,'ε')`, which is not produced by the iterator).

I’m okay with this PR if we add two warnings to the documentation:
* one for `enumerate`: Indexing into `enumerate(…)` will produce different elements than iterating over it.
  In the example above, `enumerate("τεst")[3]` will produce `(3,ε)` while iteration over `enumerate("τεst")` will produce `(2,ε)` and `(3,s)`.
  Alternatively, indexing could be restricted to something like `Enumerate{<:AbstractArray}`.
* one for `Threads.@threads`: The `:static` scheduler (which is currently the only one available) rewrites the loop to use indexing instead of iteration, so it will give unexpected results when indexing and iteration of the looped-over object produce different values.
  Technically, this has nothing to do with this PR, but it is the motivation for the original issue. Maybe we should add this warning regardless of whether we merge this PR or not.